packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -728,8 +747,17 @@ export class Bucket extends BucketBase {
 
     this.versioned = props.versioned;
     this.encryptionKey = encryptionKey;
-    this.bucketArn = resource.bucketArn;
-    this.bucketName = resource.bucketName;
+    this.bucketArn = new cdk.CrossAccountRegionPhysicalArnToken(resource.bucketArn, new cdk.Token(() => this.node.stack.formatArn({
```

Comment:
We need to work on this API's ergonomics. It's going to be something we expect every L2 to have and it's too verbose and the cognitive load is too great.

Maybe we can provide a class `AmazonResourceName` that has the cross-environment logic baked into it:

```ts
const resourceName = new cdk.AmazonResourceName(this, {
  service: 's3',
  userDefined: props.bucketName // undefined if `bucketName` is not defined, so that's easy
  // ...api here should probably be a little more elaborate to support different types of rendering the resource name component of the ARN (separators, etc)
});

this.bucketArn = resourceName.arn;
this.bucketName = resourceName.resource;
```

The `arn` and `resource` properties will return the magical tokens.

I am also curious how this could be used to implement the various `ImportXxx` heuristics where users can pass in only a name and/or ARN of resources and this class can implement the heuristics

Reply:
Agree the API is not great for Construct Library authors at the moment. Will iterate using your ideas.

I can't see how this ties to inferring the names and ARNs during importing, but perhaps we'll get some clarity when we iterate on this API a little more.

Reply:
Changed a lot. Let me know how you like the new version.